### ISSUE TITLE
Request: full codebase readiness review for GRPO training pipeline

### ISSUE THREAD

**Issue by chrisjcc** [chrisjcc/text-to-sql-coder#8]
**Title:** Request: full codebase readiness review for GRPO training pipeline — @claude please verify

**Assignee:** @claude

---

Hi @claude — please do a careful, end-to-end review of this repository with the explicit goal of verifying that the training pipeline will run (or can be made to run) for a real GRPO training job. Treat this as a production-readiness check: we need confirmation that the scripts, modules, and integrations will execute, plus a short, prioritized list of fixes or follow-up PRs if anything breaks.

Below is the scope, concrete steps to reproduce, focused checks, and expected deliverables.

---

## Scope / Files to review
Please inspect the following code paths and supporting modules:

- `scripts/train.py`  
- `scripts/evaluate.py`  
- `src/data/dataset_loader.py`  
- `src/models/model_loader.py`  
- `src/utils/hub_utils.py`  
- `src/rewards/base_reward.py`  
- `src/rewards/sql_executor.py`  
- `src/rewards/sql_reward.py`  
- `src/training/trainer.py`  
- `src/training/environment.py`  
- `config/training_config.py`

---

## What I need from you
1. Verify the pipeline runs end-to-end (dry-run) on a developer machine and report all runtime errors.  
2. Confirm the training script can be executed safely with small settings. If it cannot, provide minimal, concrete fixes or a PR.  
3. Produce a checklist of failing items grouped by severity (critical, high, medium, low), with suggested fixes and, where possible, one-line code patches or exact commands to reproduce.  
4. If tests are needed, add small runnable commands or unit tests that would catch the issues you find.

Deliverable: a short written report (markdown) with the checklist, reproduction steps, failing stack traces, and proposed fixes. If you create PRs, link them in the report.

---

## Environment and quick setup (suggested)
Use Linux or macOS if possible. There are known Unix-specific calls (signals) in the code that may not work on Windows.

Example quick setup (adjust as needed):

```bash
python -m venv .venv
source .venv/bin/activate
pip install -U pip
pip install transformers peft trl verifiers datasets huggingface_hub sqlparse wandb python-dotenv tqdm
# add any extras if errors show up (e.g., bitsandbytes, accelerate)
```

### Dry-run config suggestions
Before running a full training job, reduce sizes in config/training_config.py:
```bash
# DataConfig
train_size = 20
test_size = 4

# GRPOConfig
num_train_epochs = 1
per_device_train_batch_size = 1
gradient_accumulation_steps = 1
num_generations = 2
push_to_hub = False
```

Run the training script (dry run):
```bash
python scripts/train.py
```

If training is too heavy, run the evaluation script against a small test subset:
```bash
python scripts/evaluate.py --model_path <small_local_model_path> --num_samples 5 --num_generations 1
```

### Focused checklist (things to confirm)
**Initialization and imports:**

- Confirm sys.path usage in scripts reliably resolves project imports both from repo root and when run from CI.
- Confirm relative package imports like from ..models.model_loader import ... work when modules are executed as part of the package.

**Dependency compatibility**

- Verify functions/methods used exist on the installed package versions:
- trl.setup_chat_format
- tokenizer.apply_chat_template(...) — ensure tokenizers used have this method (it may be specific to TRL/trl-integrations).
- vf.grpo_defaults and vf.GRPOTrainer expected argument names and processing_class semantics.
- PeftModel.from_pretrained(..., torch_dtype=...) and peft_model.merge_and_unload() availability.
- huggingface_hub.repo_exists import versus HfApi().repo_exists.
- Verify bitsandbytes/quantization options used in AutoModelForCausalLM.from_pretrained are accepted by the installed transformers + bitsandbytes versions.

**Runtime correctness**

- Confirm SQLExecutor uses signal.SIGALRM safely. Note: SIGALRM is not supported on Windows. Flag cross-platform behavior and suggest a fallback.
- Check SQLExecutor._normalize_query using sqlparse: ensure sqlparse is a dependency and behaves as expected.
- Validate SchemaExtractor.extract_from_context works for the dataset schema format used.

**Tokenizer / chat format**

- tokenizer.apply_chat_template(...) is called in several places. Verify its signature matches the tokenizer installed. If not, propose a compatibility layer or wrapper.

Verifiers integration

- Verify that vf.Rubric and vf.XMLParser usage aligns with the verifiers library API and returns env objects compatible with vf.GRPOTrainer.
- Confirm the processing_class argument passed to vf.GRPOTrainer is correct (the wrapper currently passes the tokenizer object).

**LoRA / PEFT lifecycle**

- Confirm ModelLoader.get_lora_config() produces a LoraConfig object compatible with how vf.GRPOTrainer expects peft_config.
- In merge_and_save, verify the sequence: unload trainer, load base model with quantization config, PeftModel.from_pretrained, merge_and_unload, and save_pretrained works for both safetensors and state dicts.

**Hub upload**

- Validate hub_utils.upload_folder and upload_file flows with current huggingface_hub versions. Check repo_exists import path.

**Caching and memory**

- Confirm cache keys and cache size in CachedReward and SQLExecutor are safe for long runs. Verify caches are cleared appropriately (or provide TTL suggestion).
- Confirm memory cleanup sequences in merge_and_save actually free GPU memory in practice.

**Error handling and logging**

- Confirm exceptions are logged and training does not silently swallow fatal errors. For example BaseReward.__call__ returns -1.0 on exceptions which is fine, but log messages should include tracebacks for debugging.

**Platform, performance, and safety**

- Confirm operations that rely on native modules (bitsandbytes, flash_attn) degrade gracefully if not installed.
- Verify that SQLExecutor imposes reasonable result size limits and that executescript(schema) cannot be abused with large inputs.

**Smoke test commands to run**

Unit smoke test for dataset loader:

- python -c "from config.training_config import Config; from src.data.dataset_loader import SQLDatasetLoader; c = Config(); dl = SQLDatasetLoader(c.data); d=dl.load(); print(len(d['train']))"

**Dry-run training:**

# after reducing sizes in config
python scripts/train.py

Dry-run evaluation:

python scripts/evaluate.py --num_samples 5 --num_generations 1 --model_path <some small model>

Run only reward functions on a single sample:

# quick script
from config.training_config import Config
from src.rewards.sql_reward import create_sql_reward
c = Config()
reward = create_sql_reward(c.reward, 'combined')
print(reward(prompt="..", completion="...", answer="..."))

Potential known pitfalls to call out explicitly

signal.SIGALRM use is Unix specific. This will break on Windows.

tokenizer.apply_chat_template may not be present on plain AutoTokenizer. Confirm the code expects the TRL-specific tokenizer wrapper.

vf.GRPOTrainer(... processing_class=self.tokenizer ...) may be incorrect. processing_class usually expects a class or processor, not the tokenizer instance. Verify verifiers docs.

PeftModel.merge_and_unload() and from_pretrained(..., torch_dtype=...) are version-dependent. Confirm PEFT version compatibility.

huggingface_hub.repo_exists import path changes across versions. Prefer using HfApi().repo_exists if needed.

Quantization related args being passed to from_pretrained must match your transformers + bitsandbytes versions.

Expected outcome / acceptance criteria

A short report from you that includes:

Which commands you ran and on what OS/hardware.

A pass/fail for each smoke test.

Any runtime errors and stack traces.

A prioritized list of actionable fixes and suggested code patches or PRs. If fixes are trivial, include the patch or a branch/PR that applies them.

Confirmation that train and evaluate can be executed with small settings, or a clear explanation why not and next steps.

If you find no critical blockers, confirm the best practice changes we should still apply before running a long training job (dependency pinning, additional checks, CI tests, environment recommendations).

### Priority

High. This must be validated before we run larger scale experiments.

If anything is unclear, run the dry-run with minimal settings and report the first failing error. Thank you — looking forward to your findings.
